src/instrumenter/annotations.ts

```diff
@@ -546,6 +608,163 @@ export function gatherContractAnnotations(
+                 */
+                const targets = resolveAny(name, scope, compilerVersion, true);
+                const target = [...targets][0];
+
+                if (target instanceof FunctionDefinition && args.length > 0) {
+                    const params = target.vParameters.vParameters;
+
+                    mapArgsToParams(signature, args, params, localAliases);
+                }
+
+                for (const { expression, message } of properties) {
+                    const annotation = parseAnnotation(expression, target, compilerVersion);
+
+                    assert(
+                        annotation instanceof SProperty,
+                        `Only properties are allowed to be defined in macros. Unsupported annotation: ${expression}`
+                    );
+
+                    /**
+                     * Update variable names with supplied values from macro instruction.
+                     */
+                    annotation.expression.walk(walker);
+                    annotation.label = message;
+                    annotation.prefix = "#";
+
+                    /**
+                     * @todo This is temporary hack. Need to find a better way to do this.
+                     */
+                    const dummyDoc = new StructuredDocumentation(
+                        0,
+                        "0:0:0",
+                        "StructuredDocumentation",
+                        annotation.prefix + annotation.pp()
+                    );
+
+                    const metaToInject = new PropertyMetaData(
+                        dummyDoc,
+                        target,
+                        dummyDoc.text,
+                        annotation,
+                        0,
+                        dummyDoc.text
+                    );
+
+                    validateAnnotation(target, metaToInject);
+
+                    const metasToInject = injections.get(target);
+
+                    if (metasToInject === undefined) {
+                        injections.set(target, [metaToInject]);
+                    } else {
+                        metasToInject.push(metaToInject);
+                    }
+                }
+            }
+
+            /**
+             * Remove macro annotation as it will not pass type-checking.
+             * It is now represented by other annotations in `injections`,
+             * so there is no further use for macro annotation itself.
+             */
+            metas.splice(m, 1);
+        }
+    }
+
+    /**
+     * Merge injected annotations with original annotations.
+     */
+    for (const [target, metasToInject] of injections) {
+        const metas = annotationMap.get(target);
+
+        if (metas === undefined) {
+            annotationMap.set(target, metasToInject);
+        } else {
+            metas.push(...metasToInject);
+        }
+    }
+
+    /**
+     * @todo list:
+     * [TODO]   1. Source locations are needed for property maps.
```

Comment:
So this is a really tricky problem. Whereas ast nodes in normal annotations have a single source location (wherever the original annotation in the docstring is), ast nodes in instantiated macros actually have **TWO** locations:

1. The source location corresponding to the AST node inside the annotation in the yaml file
2. The location of the actual macro invocation on the target contract.

And when we want to report errors, ideally we want to show the users something like:

Error on line X col Y in file macros.yaml: Error unknown identifier 'x' in /// invaraiant x>0;
Instantiated on line Z col W in file target.sol: /// #ERC20( .... )

To support this cleanly we could do something like change the type of `src` in `SNode` to something weird like `string | MacroSrc` where `MacroSrc` contains both sources. However this is probably going to be a very invasive change.

Reply:
If we introduce macros as an **experimental feature**, we can try to map produced annotations to macro annotation that is present in the source. It would be very broad, still it will not block merging. We should have a disclaimer that should explicitly state that "_implementation is still not in final state_".

Reply:
Fixed with 072f378